python/flatbuffers/builder.py

```diff
@@ -442,6 +443,38 @@ def CreateByteVector(self, x):
 
         return self.EndVector(len(x))
 
+    def CreateNumpyVector(self, x):
+        """CreateNumpyVector writes a numpy array into the buffer."""
+
+        if np is None:
+            # Numpy is required for this feature
+            raise NumpyRequiredForThisFeature("Numpy was not found.")
+
+        if not isinstance(x, np.ndarray):
+            raise TypeError("non-numpy-ndarray passed to CreateNumpyVector")
+
+        if x.data.ndim > 1:
+            raise TypeError("multidimensional-ndarray passed to CreateNumpyVector")
+
+        self.StartVector(x.itemsize, x.size, x.dtype.alignment)
+
+        # Ensure little endian byte ordering
+        if x.dtype.str[0] == "<":
+            x_lend = x
+        else:
+            x_lend = x.byteswap()
```

Comment:
So my intuition for `byteswap` was to avoid looping over a long vector in python. `numpy` uses the C for-loop and the cpu `BSWAP` [instructions](https://github.com/numpy/numpy/blob/083aedba2f49e1b490d3e122f5927f0718cf202c/numpy/core/src/multiarray/ctors.c#L333) (if available), which I assumed to be much quicker than anything available in python. Additionally `byteswap` deals with the datatype specifics (different swapping instructions depending on `itemsize`). The additional cost of a memory allocation should be small compared to the overhead of the python for loop.

I would definitely try to avoid modification of the input data. The `in_place` default is `False`, which should prevent that, however, I think enforcing it, just in case they change it in future versions, should be done.

To verify my assumptions I did a quick benchmark for three different variants of the function.

1. current implementation
2. Python for loop with the correct `builder.Prepend` method
3. Python for loop with a direct call to `struct.pack_into`

I serialize vectors of different sizes 1000 times and compute average and maximum time needed for serialization.
[Code](https://github.com/felixfrank/flatbuffers/blob/master/my_tests/CreateVectorBenchmark.py)
Results:
![createvectorbenchmark_results](https://user-images.githubusercontent.com/16225194/43355077-b694e6c8-9256-11e8-9627-ac5a667dbd56.png)

Except for very small vectors `byteswap` significantly outperforms the other two methods. Note that additionally the other two methods would need big `if` clauses to modify the behavior according to the correct datatype.